Added a validator for Read-Only files.

FelixGV opened the pull request:
Given a data directory, the ```ReadOnlyFileValidator``` class will check a few things:
- Index files are well formed:
  - Well sorted
  - no negative offset
- Data files are well formed:
  - Number of records for a given key hash is > 0
  - Key and value sizes are > 0
- Index and data files match together:
  - No gaps in data files, when traversing them with the offsets from index files
  - Raw keys in data files match the hashes found in their index files

It then prints a summary of the amount and kinds of problems found.

I am posting this here in case anyone is curious, but I might modify it further before being ready to merge it in, so please don't merge it in yet yourselves. Thanks.

arunthirupathi commented:
Do you think a check to read .metadata file and make sure the checksum matches between the data and index file is worth adding ?

That will help in identifying the disk corruption.

The way I see it is , these checks if errors out we have errors. But absence of errors does not mean they are good.

arunthirupathi commented on src/java/voldemort/store/readonly/ReadOnlyFileValidator.java:
```diff
@@ -0,0 +1,363 @@
+                problemsDetectedInIndexFiles += validateIndexFile(file, verbose);
+                problemsDetectedInIndexAndDataFilesMatch += validateIndexAndDataFileMatch(file, verbose);
+            } else if (file.getName().endsWith(ReadOnlyUtils.DATA_FILE_EXTENSION)) {
+                problemsDetectedInDataFiles += validateDataFile(file, verbose);
```
Should you check for existence of .index file here ?

arunthirupathi commented on src/java/voldemort/store/readonly/ReadOnlyFileValidator.java:
```diff
@@ -0,0 +1,363 @@
+            return 0;
+        } else if (cmp == 0) {
+            // The same key hash or offset appeared twice in a row. Unexpected!
+            System.err.println("We found two identical consecutive key " + description +
```
Should we print them only under verbose conditions ?

FelixGV replied:
Yes, good point.

arunthirupathi commented on src/java/voldemort/store/readonly/ReadOnlyFileValidator.java:
```diff
@@ -0,0 +1,363 @@
+            }
+
+            if (keySize == 0) {
+                System.out.println("Key size 0!");
```
Same verbose here, also this uses System.out other uses System.err

FelixGV replied:
Good catch, I'll refactor all of that.

FelixGV commented:
Thanks for the review @arunthirupathi. I amended my previous commit.

I added a bash script to easily run the ```ReadOnlyFileValidator```. I also added disclaimers about what this script does not do.

I'm good to go as far as merging this in.

arunthirupathi commented:
Looks good to me
